lib/teleterm/gateway/gateway.go

```diff
@@ -172,6 +196,39 @@ func (g *Gateway) LocalPort() string {
 	return g.cfg.LocalPort
 }
 
+// SetLocalPort attempts to create a listener on the specified port. If successful, it stops
+// g.localProxy and starts a new one with the new listener and then updates the fields on gateway.
+//
+// If it fails it's imperative that the current proxy is kept intact and the fields on gateway are
+// not changed. This way if the user attempts to change the port to one that cannot be obtained,
+// they're able to correct that mistake and choose a different port.
+func (g *Gateway) SetLocalPort(port string) error {
+	result, err := newListenerAndLocalProxy(*g.cfg, port)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	previousPort := g.cfg.LocalPort
+
+	g.closeCancel()
+	if err = g.localProxy.Close(); err != nil {
+		g.cfg.Log.WithError(err).Warnf("Failed to close the previous proxy on port %s.", previousPort)
+	}
```

Comment:
Is failing to close a listener a real concern usually?

Of course the logic in `alpn.LocalProxy.Close()` might grow and include other stuff than just closing the listener. OTOH if `SetLocalPort` returned an error in this scenario, I guess we'd need to close the new listener then. Maybe that would be a behavior that could be described as more correct?